peer/pendingheap/heap.go

```diff
@@ -0,0 +1,187 @@
+// Copyright (c) 2017 Uber Technologies, Inc.
+//
+// Permission is hereby granted, free of charge, to any person obtaining a copy
+// of this software and associated documentation files (the "Software"), to deal
+// in the Software without restriction, including without limitation the rights
+// to use, copy, modify, merge, publish, distribute, sublicense, and/or sell
+// copies of the Software, and to permit persons to whom the Software is
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Software.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package pendingheap
+
+import (
+	"container/heap"
+	"context"
+	"fmt"
+	"sync"
+
+	"go.uber.org/yarpc/api/peer"
+	"go.uber.org/yarpc/api/transport"
+)
+
+type pendingHeap struct {
+	lock      sync.RWMutex
+	transport peer.Transport
+	peers     []*peerScore
+
+	// next is an incrementing counter for every push, which is compared when
+	// scores are equal. This ends up implementing round-robin when scores are
+	// equal.
+	next int
+}
+
+func (ph *pendingHeap) Choose(ctx context.Context, req *transport.Request) peer.StatusPeer {
+	ph.lock.Lock()
+	ps, ok := ph.popPeer()
+	ph.lock.Unlock()
+	if !ok {
+		return nil
+	}
+
+	ph.lock.Lock()
+	// Note: We push the peer back to reset the "next" counter.
+	// This gives us round-robin behavior.
+	ph.pushPeer(ps)
+	ph.lock.Unlock()
+
+	return ps.peer
+}
+
+func (ph *pendingHeap) Add(p peer.StatusPeer) peer.Subscriber {
+	ps := &peerScore{peer: p, heap: ph}
+	ps.score = scorePeer(p)
+
+	ph.lock.Lock()
+	ph.pushPeer(ps)
+	ph.lock.Unlock()
+	return ps
+}
+
+func (ph *pendingHeap) Remove(p peer.StatusPeer, sub peer.Subscriber) {
+	ps := sub.(*peerScore)
```

Comment:
do we have a guarantee that this will always be of the right type? want to make sure we don't end up causing a panic in prod, might be safer to do a check?